src/EventStore.Core/GossipStatsCollector.cs

```diff
@@ -0,0 +1,110 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+
+namespace EventStore.Core;
+
+public readonly struct GossipStats {
+	public string Name { get; init; }
+	public double MovingAverage { get; init; }
+	public double Max { get; init; }
+	public double Min { get; init; }
+	public double Last { get; init; }
+
+	internal GossipStats(string name, double movingAverage, double max, double min, double last) {
+		Name = name;
+		MovingAverage = movingAverage;
+		Max = max;
+		Min = min;
+		Last = last;
+	}
+}
+
+internal class GossipRequestActivity : IDisposable {
+	private readonly Stopwatch _stopwatch;
+	private readonly GossipStatsCollector _collector;
+
+	internal GossipRequestActivity(GossipStatsCollector collector) {
+		_collector = collector;
+		_stopwatch = Stopwatch.StartNew();
+	}
+
+	public void Dispose() {
+		_collector.Report(_stopwatch.Elapsed);
+	}
+}
+
+/// <summary>
+/// Computes a Simple Moving Average of all the gossip read calls.
+/// https://en.wikipedia.org/wiki/Moving_average
+/// </summary>
+public class GossipStatsCollector {
+	private double _sma = 0;
+	private double _last = 0;
+	private double _max = 0;
+	private double _min = 0;
+	private int _dataPointCount = 0;
+	private readonly object _lock = new object();
+
+	public string Name { get; init; }
+
+	internal GossipStatsCollector(string name) {
+		Name = name;
+	}
+
+	internal void Report(TimeSpan duration) {
+		lock (_lock) {
+			if (_dataPointCount == 0) {
+				++_dataPointCount;
+				_sma = duration.TotalMilliseconds;
+				_last = _sma;
+				_max = _sma;
+				_min = _sma;
+				return;
+			}
+
+			_sma += 1/((double) _dataPointCount) * (duration.TotalMilliseconds - _last);
+			_max = Math.Max(_max, duration.TotalMilliseconds);
+			_min = Math.Min(_min, duration.TotalMilliseconds);
+			_last = duration.TotalMilliseconds;
+			++_dataPointCount;
+		}
+	}
+
+	public IDisposable StartRequest() {
+		return new GossipRequestActivity(this);
+	}
+
+	internal GossipStats Collect() {
+		GossipStats stats;
+
+		lock (_lock) {
+			stats = new GossipStats(Name, _sma, _max, _min, _last);
+		}
+
+		return stats;
+	}
+}
+
+public class GossipStatsCollectors {
+	public static readonly GossipStatsCollectors Default = new GossipStatsCollectors();
+
+	private readonly ConcurrentDictionary<string, GossipStatsCollector> _collectors =
+		new ConcurrentDictionary<string, GossipStatsCollector>();
+
+	private GossipStatsCollectors() {}
+
+	public GossipStatsCollector New(string name) {
```

Comment:
Minor nitpick, this method should be called `Create` since `new` is for constructors